signer/src/schnorr.rs

```diff
@@ -0,0 +1,156 @@
+        assert!(domain_string.len() <= 20);
+        domain_string = &domain_string[..std::cmp::min(domain_string.len(), 20)];
+        let mut bytes = format!("{:*<20}", domain_string).as_bytes().to_vec();
+        bytes.resize(32, 0);
+
+        bytes
+    }
+
+    // This function uses a cryptographic hash function to create a uniformly and
+    // randomly distributed nonce.  It is crucial for security that no two different
+    // messages share the same nonce.
+    fn blinding_hash<H>(&self, kp: &Keypair, input: H) -> ScalarField
+    where
+        H: Hashable,
+    {
+        let mut hasher = VarBlake2b::new(32).unwrap();
+
+        let mut roi: ROInput = input.to_roinput();
+        roi.append_field(kp.public.to_point().x);
+        roi.append_field(kp.public.to_point().y);
+        roi.append_scalar(kp.secret.to_scalar());
+        roi.append_bytes(&[self.network_id.into()]);
+
+        hasher.update(roi.to_bytes());
+
+        let mut bytes = [0; 32];
+        hasher.finalize_variable(|out| bytes.copy_from_slice(out));
+        // Drop the top two bits to convert into a scalar field element
+        //   N.B. Since the order p is very close to 2^m for some m, truncating only creates
+        //   a tiny amount of bias that should be insignificant and keeps the implementation
+        //   simple by avoiding reduction modulo p.
+        bytes[bytes.len() - 1] &= 0b0011_1111;
+
+        ScalarField::from_random_bytes(&bytes[..]).expect("failed to create scalar from bytes")
+    }
+
+    // This function uses a cryptographic hash function (based on a sponge construction) to
+    // convert the message to be signed (and some other information) into a uniformly and
+    // randomly distributed scalar field element.  It uses Mina's variant of the Poseidon
+    // SNARK-friendly cryptographic hash function.
+    // Details: <https://github.com/o1-labs/cryptography-rfcs/blob/httpsnapps-notary-signatures/mina/001-poseidon-sponge.md>
+    fn message_hash<S>(&mut self, pub_key: &PubKey, rx: BaseField, input: S) -> ScalarField
+    where
+        S: Signable,
+    {
+        let mut roi: ROInput = input.to_roinput();
+        roi.append_field(pub_key.to_point().x);
+        roi.append_field(pub_key.to_point().y);
+        roi.append_field(rx);
+
+        // Set sponge initial state (explicitly init state so signer context can be reused)
+        // N.B. Mina sets the sponge's initial state by hashing the input's domain bytes
+        self.sponge.state = vec![BaseField::zero(); self.sponge.state.len()];
+        self.sponge.sponge_state = SpongeState::Absorbed(0);
+        self.sponge
+            .absorb(&[
+                BaseField::from_bytes(&Schnorr::<SC>::domain_bytes::<S>(self.network_id))
+                    .expect("invalid domain bytes"),
+            ]);
+        self.sponge.squeeze();
+
+        // Absorb random oracle input
+        self.sponge.absorb(&roi.to_fields());
```

Comment:
do you think it'd make sence to have `to_fields()` consume `roi` so you can only call it once?

Reply:
I really don't like the ROI approach btw because it feels prone to misuse, it looks like we're just shoving bits and field elements together and there might be collisions or ambiguity in the way we do this.

I think a better design would do this:

* absorb a field just absorbs a field element
* if you absorb bits/bytes, it caches them in a queue (and also prepends the length of the bits/bytes every time you call these functions in the queue as well)
* if you absorb bits/bytes when the queue gets bigger than the size of a field element, you convert some of the queue to a field element
* if you absorb field (or call `to_fields()`) when the bits queue is not empty, then you first convert your bit queue into field, stick that into the field array, and then absorb your field

Reply:
I agree with your suggestions, but I think this should be done in another ticket.  The reason is because if we cannot make changes to this without breaking compatibility with mainnet mina transaction signatures, which is one of my main objectives (to be the same as the OCaml code).  That said, these are very good suggestions -- especially the length-padding part.  Thus, I've opened up a separate ticket for improving and simplifying the ROI structure and length padding things.  https://github.com/o1-labs/proof-systems/issues/244